### ISSUE TITLE
Preserve Org link search options when resolving roam: links

### ISSUE THREAD

**Issue by FrederickGeek8** [org-roam/org-roam#2480]
## Brief Abstract
In short, I would like the automatic expansion of `roam:` to `id:` links to preserve links to headings using the existing `id:_::*Heading` org-mode link format (you can see an example [in the first table in this `org-mode` docs page](https://orgmode.org/manual/External-Links.html#External-links)). 

This functionality would exist for a situation where I **do not** want to register a dedicated `ID` for that specific heading via `org-store-link`. This can be useful for one-off links or non-essential headings, such as `[[roam:Complicated Topic::*Definition]]`, and regardless can be quicker to use as a shorthand then storing a non-essential heading link.

### Existing functionality in Org-Mode and Org-Roam v2
**(Org-_Roam_) Resolution of `roam:` to `id:` links:**
When quickly writing documents, I will often use the `[[roam:Page]]` shorthand to link to org-roam (v2) documents. On save, these links will be resolved to `id:` links on save via [`org-roam-link-replace-all`](https://github.com/org-roam/org-roam/blob/2a630476b3d49d7106f582e7f62b515c62430714/org-roam-node.el#L762). (Currently `org-roam-link-replace-all` delegates to [`org-roam-link-replace-at-point`](https://github.com/org-roam/org-roam/blob/2a630476b3d49d7106f582e7f62b515c62430714/org-roam-node.el#L741))

**(Org-_Mode_) Headline search in `id:` links:**
`org-mode` natively supports "headline search" on `id` links [(as described in the first table in this `org-mode` doc page)](https://orgmode.org/manual/External-Links.html#External-links). The result is that when a `[[id:BLA-LBLA-BLALB::*My Heading][Some Page]]` linked is click, the opened document will jump to the heading "My Heading" in the linked org document.

### Current behavior in Org-Roam (to be changed)
**Resolution of `roam:_::*heading` to `id:_::*heading` links:**
Currently resolution of `[[roam:Page::*Heading]]` links is not _attempted_ by [`org-roam-link-replace-all`](https://github.com/org-roam/org-roam/blob/2a630476b3d49d7106f582e7f62b515c62430714/org-roam-node.el#L762) -- the `[[roam:Page::*Heading]]` link will remain in the document unchanged.

## Proposed functionality
Ideally, [`org-roam-link-replace-at-point`](https://github.com/org-roam/org-roam/blob/2a630476b3d49d7106f582e7f62b515c62430714/org-roam-node.el#L741) would resolve
`[[roam:Page::*heading]]`
to
`[[id:MYID-MYID-MYID::*Heading][Page::*Heading]]`

My intent is that **at the very least** resolution of the `roam:` link would preserve the `::*Heading` and place it in the `id:` link. Ideally (for me) it would also preserve the `::*Heading` in the "description" as well (e.g. `[Page::*Heading]`) for clarity.

## Proposed Implementation
I imagine that this functionality should be a relatively simple addition to [`org-roam-link-replace-at-point`](https://github.com/org-roam/org-roam/blob/2a630476b3d49d7106f582e7f62b515c62430714/org-roam-node.el#L741). I can _maybe_ submit a PR for this (eventually), but I don't have a good development environment set up right now, and I'm not familiar with contribution procedures here (yet). Besides, more knowledgeable people/contributor may have opinions on the correct handling of this.

## Please check the following:

- [X] No similar feature requests
    - This is somewhat related to an earlier (implemented) feature request #548 

**A small note:**
My brain is a bit jumbled today so apologies if this Issue is as jumbled as it appears to me.

Also,
Thanks!

**Comment by FrederickGeek8** [comment]
I have more time now to try and work on this. Would a change like this be accepted/acceptable?

**Comment by FrederickGeek8** [comment]
> Please test the following and let me know

Your snippet seems to work! I just tested by renaming it to `org-roam-link-replace-at-point-test` and executing it over a link of the form `[[roam:Page::*Heading]]` via `M-:`, and it worked as desired -- resolving to a form of `[[id:MYID-MYID-MYID::*Heading][Page::*Heading]]`!

**Comment by FrederickGeek8** [comment]
I meant "integrated" as in getting this change moved into the `org-roam` library proper, so your latter option. You had referenced a in-progress (now merged) PR (that I did not quite understand the relevance of) [in your comment above](https://github.com/org-roam/org-roam/issues/2480#issuecomment-2570625178), and so I'm not sure if there is more going on behind the scenes needs to be done to support making this change.

Hopefully Dustin has some advice on how to proceed!

**Comment by hpfr** [comment]
IMO, if you are linking to a heading, it’s relevant enough to be its own node in the graph. But this seems like it shouldn’t be difficult to support.

Re: case insensitivity, I think it might be better to implement `:complete` in `org-link-set-parameters` for the `roam` link type so that `org-insert-link` suggests nodes, and then users’ completion stacks handle case insensitivity. We already have completion at point so that should be reusable.

**Comment by akashpal-21** [comment]
@hpfr You are correct. Case insensitivity is much better handled by auto completion stack, so the patch I provided is only applicable to me because I do not use case insensitivity in my autocompletion framework, there is no rationale - but since 
`org-roam-node-from-title-or-alias` already is equipped to handle case sensitivity in its second argument - I figured why not supply a variable to control this ? This allows me to keep case sensitivity in autocompletion to work independently to org-roam-link resolver.

In my personal case - I use org-roam as a singular point of entry to all my notes - this includes some very big documents - book size entries that have a labrynth of underling headings - sometimes I want to link to a particular heading - but I use ID for the page only - and not headings so link-search becomes indispensable in linking.

One time, someone wanted to use roam links to pre emptively create links to future pages that are not yet created - my idea of using roam links emerged from there -  but without my underlying infrastructure in dealing with roam links independently they are merely dangling links, that should be converted to id links without delay - but in my case roam links provide a secondary layer of traversal independently of id links - I do not know how useful it is to anyone else - even to me ;)

**Comment by akashpal-21** [comment]
But case-insensitivity here could be important too - consider the case -

user wants to create a future page "FUTURE PAGE" but user may not know how they would capitalise it right now - so they might use [[roam:future page]] and this would work in the future too - when the page is created and conversion takes place to id. But without case insensitivity - [[roam:future page]] would not convert to [[id:value>][FUTURE PAGE]]

**Comment by hpfr** [comment]
Great point. With future links in mind, I think case insensitive roam: link resolution should be the default (and not even exposed as an option, since I don’t think allowing node titles that differ only by case is viable UX).
